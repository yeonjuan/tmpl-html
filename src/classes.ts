import {isStringLike, } from "./utils";

type Value = string | number | boolean | undefined | null;
type RecordValue = Record<string, unknown>;
type ArrayValue = Arg[];
type Arg = Value | RecordValue | ArrayValue;

const hawOwn = {}.hasOwnProperty;

export default function classes(...args: Arg[]): string {
  const classnames: string[] = [];

  for (let i = 0; i < args.length; i++) {
    const arg = args[i];
    if (isStringLike(arg)) {
      classnames.push(String(arg));
    } else if (Array.isArray(arg)) {
      arg && classes(arg);
    } else if (typeof arg === 'object' && !!arg) {
      for (const key in arg) {
        (arg[key] && hawOwn.call(arg, key) && classnames.push(key))
      }
    }
  }
  return `"${classnames.join(" ")}"`;
}
