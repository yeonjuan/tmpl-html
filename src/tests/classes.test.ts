import classes from '../classes';

test.each([
  [{foo: true}],
  [{foo: false}],
  [{foo: true, bar: true}],
])('classes - record', (value) => {
  expect(classes(value)).toMatchSnapshot()
});


test.each([
  ["foo", "bar"],
  ["foo", undefined],
  ["foo", null, "bar"],
])('classes - string', (...value) => {
  expect(classes(...value)).toMatchSnapshot()
});
