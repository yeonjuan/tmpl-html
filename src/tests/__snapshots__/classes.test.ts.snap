// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`classes - record 1`] = `"\\"foo\\""`;

exports[`classes - record 2`] = `"\\"\\""`;

exports[`classes - record 3`] = `"\\"foo bar\\""`;

exports[`classes - string 1`] = `"\\"foo bar\\""`;

exports[`classes - string 2`] = `"\\"foo\\""`;

exports[`classes - string 3`] = `"\\"foo bar\\""`;
